src/DependencyInjection/DI/src/ServiceLookup/ConstantCallSite.cs

```diff
@@ -11,6 +11,11 @@ internal class ConstantCallSite : ServiceCallSite
 
         public ConstantCallSite(Type serviceType, object defaultValue): base(ResultCache.None)
         {
+            if (defaultValue != null && !serviceType.IsInstanceOfType(defaultValue))
```

Comment:
It might be worth validating that the service type is reference/nullable type when the constant is null. This doesn't feel too different in principle to the validation being added.

Reply:
We don't support value type services. So I'm not sure it's worth it.